#📂 **A.A. Cálculo de Momentos: Algoritmos para el Cálculo de Momentos y Validación Computacional**

---

**GRUPO G**

### 👥 **Estudiantes:**
* Emerson Sebastian Chamba Galarza
* Matias Sebastian Labanda Pineda
* Lenin Fabricio Macas Cabrera
* Pilar Valentina Naranjo Quizhpe
* Mateo Sebastian Pucha Carrera


###📊 **Desarrollo Algorítmico**

In [ ]:
import pandas as pd
import numpy as np

# 1. Cargar dataset
df = pd.read_csv("datos_loja.csv")

# 2. Variable
datos = df["Sin_Alcantarillado"].values


# 3. Esperanza (media) manual
def calcular_esperanza(datos):
    suma = 0
    n = 0

    for x in datos:
        suma = suma + x
        n = n + 1

    return suma / n


# 4. Varianza manual
def calcular_varianza(datos, insesgada=True):
    media = calcular_esperanza(datos)

    suma = 0
    n = 0

    for x in datos:
        suma = suma + (x - media) ** 2
        n = n + 1

    if insesgada:
        return suma / (n - 1)
    else:
        return suma / n


# 5. Cálculo manual
media_manual = calcular_esperanza(datos)
var_manual = calcular_varianza(datos, insesgada=True)

# 6. Validación con NumPy
media_numpy = np.mean(datos)
var_numpy = np.var(datos, ddof=1)

# 7. Resultados
print("=== RESULTADOS ===\n")

print(f"Media manual: {media_manual}")
print(f"Media NumPy:  {media_numpy}\n")

print(f"Varianza manual: {var_manual}")
print(f"Varianza NumPy:  {var_numpy}\n")

print("Diferencias:")
print(f"Media: {abs(media_manual - media_numpy)}")
print(f"Varianza: {abs(var_manual - var_numpy)}")

=== RESULTADOS ===

Media manual: 2382.625
Media NumPy:  2382.625

Varianza manual: 5203416.916666667
Varianza NumPy:  5203416.916666667

Diferencias:
Media: 0.0
Varianza: 0.0


### 📊 **Comparación de Resultados**
---

| Método | Media     | Varianza        |
|--------|----------|----------------|
| Manual | 2382.625 | 5203416.9167|
| NumPy  | 2382.625 | 5203416.9167 |


Al comparar los valores obtenidos mediante el cálculo manual y los generados por NumPy, se observa que tanto la media como la varianza presentan resultados exactamente iguales.

No existe diferencia en los decimales, lo que indica que la implementación manual es correcta y que no hay errores de precisión de punto flotante en este caso.

###🧠 **Análisis de Resultados**

Es importante que un ingeniero en computación entienda cómo se calculan la media y la varianza, en lugar de depender completamente de librerías como NumPy. Esto permite interpretar correctamente los resultados y detectar posibles errores en los datos.

Además, implementar estos cálculos manualmente fortalece el pensamiento lógico y ayuda a comprender conceptos clave como la diferencia entre población y muestra.

Aunque las librerías son eficientes, pueden usarse incorrectamente si no se conocen sus parámetros. Por ello, entender su funcionamiento garantiza un uso adecuado y resultados confiables.
Esto demuestra que los resultados computacionales deben ser siempre validados para garantizar su confiabilidad.